Added code for rsync added in tasks

ramkiperiy opened the pull request:
Adding new code for rsync task. This task can run parallel along with any other tasks.
it does snap creation and rsync the data to fs mount then removal of snap and data dir,
this steps will be done in loop till parallel task execution completes
Code can support for workunit task in parallel
Updated code based on review comments
https://github.com/ceph/ceph-qa-suite/pull/1241

Signed-off-by: ramkiperiy <rperiyas@redhat.com>

ramkiperiy committed:
Added code for rsync added in tasks

Adding new code for rsync task. This task can run parallel along with any other tasks.
it does snap creation and rsync the data to fs mount then removal of snap and data dir,
this steps will be done in loop till parallel task execution completes
Code can support for workunit task in parallel
Updated code based on review comments
https://github.com/ceph/ceph-qa-suite/pull/1241

Signed-off-by: ramkiperiy <rperiyas@redhat.com>

ceph-jenkins commented:
Can one of the admins verify this patch?

jcsp commented:
add to whitelist

jcsp commented:
ok to test

ramkiperiy commented:
Attaching passed test logs (teuthology.log & config.yaml).
Scenarios:
      1. normal rsync test without snap
      2. rsync with snap enabled 
      3. rsync with workunit task in parallel.
[normal_rsync.yaml.zip](https://github.com/ceph/ceph-qa-suite/files/594468/normal_rsync.yaml.zip)
[normal_rsync_teuthology.log.zip](https://github.com/ceph/ceph-qa-suite/files/594469/normal_rsync_teuthology.log.zip)
[rsync_workunit.yaml.zip](https://github.com/ceph/ceph-qa-suite/files/594470/rsync_workunit.yaml.zip)
[rsync_workunit_teuthology.log.zip](https://github.com/ceph/ceph-qa-suite/files/594472/rsync_workunit_teuthology.log.zip)
[snap_rsync.yaml.zip](https://github.com/ceph/ceph-qa-suite/files/594471/snap_rsync.yaml.zip)
[snap_rsync_teuthology.log.zip](https://github.com/ceph/ceph-qa-suite/files/594473/snap_rsync_teuthology.log.zip)

jcsp commented on tasks/rsync.py:
```diff
@@ -0,0 +1,218 @@
+
+        if self.workUnit:
+            assert self.snapEnable == False, \
+                self.logger.error('Workunit task and Snapshot enable cannot co-exist asserting rsync')
```
The second argument to assert should just be a string.  The way it's written here, if you hit this assert then the assert's message is going to be whatever the return value of logger.error is (probably None).

ramkiperiy replied:
I will update the code only with assert message (second argument)

jcsp commented on tasks/rsync.py:
```diff
@@ -0,0 +1,218 @@
+
+            if self.snapEnable:
+                # Create file and add data to the file
+                self.my_mnt.write_test_pattern("{}/file_a".format(self.sourceDir), self.fileSize * 1024 * 1024)
```
The snapshot mode and setting datadir should not be mutually exclusive -- setting snapEnable should just control whether snapshots are used, not whether a test pattern is written to the source directory.

I'm not sure we're ever going to need the mode where it's calling write_test_pattern at all -- the original purpose of this task is to run behind another workunit, not to write its own files.  If you want to keep this code, then maybe just only run it if no data directory was set.

ramkiperiy replied:
John, I tried that but mount point and workunit directory will be in testdir which is "/home/ubuntu/cephtest" and snapshot creation can be done in cephfs mount point which will not be having any data because of this used write_pattern.

Snap creation will fail in workunit directory since it is normal linux kernel directory.

Let me know you thoughts, i will change the code if required.

jcsp replied:
> Snap creation will fail in workunit directory since it is normal linux kernel directory.

If you're doing a workunit from client.0 then you'll have /home/ubuntu/cephtest/mnt.0/workunit.client.0, and you can create snapshots in /home/ubuntu/cephtest/mnt.0/workunit.client.0/.snap

jcsp commented on tasks/rsync.py:
```diff
@@ -0,0 +1,218 @@
+
+                # Create Snapshot
+                self.my_mnt.run_shell(["mkdir", "{}".format(self.snapShot)])
+                self.iteration += 1
```
Iteration does not need to be a class attribute, it could just be local to the function

ramkiperiy replied:
Even i had same idea but if we are not using snapshot scenario then it will be declared without any usage, to avoid it declared it as class attribute if snap enabled.

jcsp replied:
Still, it is always better to only declare a variable at the scope that you need it -- if it's only used inside one function, then declare it inside that function

ramkiperiy replied:
okay, i will make this change.

jcsp commented on tasks/rsync.py:
```diff
@@ -0,0 +1,218 @@
+        if self.snapEnable:
+            # Enable snapshots
+            self.fs.mon_manager.raw_cluster_cmd("mds", "set", "allow_new_snaps", "true", "--yes-i-really-mean-it")
+            self.snapDummy = self.dataDir + '/.snap/snap'
```
"snapdummy" variable is only used one place, so get rid of it and just calculate the path in-place below

ramkiperiy replied:
okay.

jcsp commented on tasks/rsync.py:
```diff
@@ -0,0 +1,218 @@
+                self.my_mnt.run_shell(["rsync", "-azvh", "{}".format(self.sourceDir), "rsyncdir/dir1/"])
+
+        except CommandFailedError:
+            self.logger.error('rsync command failed asserting rsync')
```
This exception handler seems unnecessary: we're not logging any extra information other than to say where it happened, and that will be indicated in the backtrace when the exception gets logged later.  The _run exception handler will print the backtrace.

jcsp commented on tasks/rsync.py:
```diff
@@ -0,0 +1,218 @@
+        self.logger.info(x)
+
+    def stop(self):
+        time.sleep(self.runTime)
```
Why is this sleep needed?

ramkiperiy replied:
This will be useful if the user executes rsync as sequential. for eg: if this is not set then rsync while loop iteration happens only once, so added wait time which will make sure rsync runs till runTime.

ramkiperiy replied:
yes correct without workunit. it will run rsync as normal IO's in particular mount point.

jcsp commented on tasks/rsync.py:
```diff
@@ -0,0 +1,218 @@
+            self.logger.exception("Exception in do_rsync:")
+            raise
+
+    def log(self, x):
```
This function is unused (and unnecessary)

jcsp commented on tasks/rsync.py:
```diff
@@ -0,0 +1,218 @@
+
+        # Using first mount point of cephfs for run rsync test.
+        if len(mounts):
+            self.my_mnt = mounts.pop(0)
```
Let's allow configuring which mount point to use.  Ultimately we will want to run this with two clients, and point the workunit at one and the rsync task at the other.

Currently this is possible by letting the rsync task use client.0 by default, and pointing the workunit at client.1, but it would be better to make it explicit.

ramkiperiy replied:
i will make this change.

ramkiperiy commented on tasks/rsync.py:
```diff
@@ -0,0 +1,218 @@
+        self.my_mnt.run_shell(["mkdir", "rsyncdir"])
+        self.my_mnt.run_shell(["mkdir", "rsyncdir/dir1"])
+
+        while not self.stopping.is_set():
```
Even i faced the problem but not immediately, it was at the end when workunit task completed and removed the directory and rsync was just inside while loop. tried increasing the waitTime at end of while loop not seen the problem.
may be i can change default waittime to 10s

jcsp replied:
The important thing is not to wait for a particular length of time, but to wait *until* the directory exists (i.e. keep checking with os.path.exists or similar and proceed when it exists)

jcsp replied:
The issue during the end of the task is harder, especially with snapshots: if we have created snapshots then the workunit task will fail when it tries to remove its directory.  We might need to also modify the workunit task to notice if there are snapshots when it tries to remove its directory, and make it wait for those snapshots to go away.

If necessary we could also have the rsync task expose a reference to itself on the global ctx object, and have the workunit task cooperate with it in that way.

For the non-snapshot case, I guess we probably just see the rsync operation itself failing when it tries to run while the workunit directory is removed: in that case we would need an exception handler that checks if the workunit dir is gone, and if it is gone then terminates the rsync task, and if it is still there then reraises the exception.

ramkiperiy committed:
Updated code with below changes.

1. Updated attribute names
2. Added code for if_directory_exists before rsync
3. Changes related to workunit directory structure
4. Removed unwanted functions, variables, exceptions and sleep timeouts.

Signed-off-by: ramkiperiy <rperiyas@redhat.com>

ramkiperiy commented:
Attaching test config and log file used for executing rsync data from workunit directory.
[rsync_config.yaml.zip](https://github.com/ceph/ceph-qa-suite/files/617020/rsync_config.yaml.zip)
[rsync_teuthology.log.zip](https://github.com/ceph/ceph-qa-suite/files/617021/rsync_teuthology.log.zip)

ramkiperiy committed:
Code has below changes

1. Updated code to work with workunit snap and rsync parallel, which checks event state from workunit task.

Signed-off-by: ramkiperiy <rperiyas@redhat.com>

Commit:
tasks/workunit: make 'all' an either-or thing

...vs operating on named clients.  Simplifies
the code to add observers in the next commit.

Signed-off-by: John Spray <john.spray@redhat.com>
(cherry picked from commit 4eb81e7202a47b959e6b5a0b0d4c5e6ab190d028)

Commit:
tasks/workunit: add 'observer' functionality

This is to enable tasks like the forthcoming
'rsync' task to cooperate with workunit
during setup and teardown.

Signed-off-by: John Spray <john.spray@redhat.com>
(cherry picked from commit 432a881a56f7ea9ba5432d26fbb3253bff040b91)
Signed-off-by: ramkiperiy <rperiyas@redhat.com>

Commit:
tasks/workunit: add 'observer' functionality

This is to enable tasks like the forthcoming
'rsync' task to cooperate with workunit
during setup and teardown.

Signed-off-by: John Spray <john.spray@redhat.com>
(cherry picked from commit c00ad61b6d0cfaab9e948f75bb5ee305fd510f1e)
Signed-off-by: ramkiperiy <rperiyas@redhat.com>

Commit:
tasks/workunit: add 'observer' functionality

This is to enable tasks like the forthcoming
'rsync' task to cooperate with workunit
during setup and teardown.

Signed-off-by: John Spray <john.spray@redhat.com>
(cherry picked from commit 432a881a56f7ea9ba5432d26fbb3253bff040b91)

ramkiperiy commented:
Hi John, 
    Any update on this PR ?

liewegas commented:
ceph-qa-suite is now merged into ceph.git.  Please resubmit this PR against ceph.git.
